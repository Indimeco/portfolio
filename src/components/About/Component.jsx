import React from 'react';
import Rotary from './Rotary';
import Log from './Log';

export const About = () => {
	return (
		<div data-test="about">
			<Rotary />
			<Log />
		</div>
	);
};

export default About;
