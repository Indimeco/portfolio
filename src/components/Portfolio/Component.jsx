import React, { useState } from 'react';
import {
	PortfolioList,
	PortfolioItem,
	PortfolioPreview,
	PortfolioControl,
	Thumbnail,
} from './Component.style';
import { data } from './data';

const PortfolioEntry = ({ item, isActive, onChange }) => (
	<PortfolioItem>
		<PortfolioPreview>
			<Thumbnail>
				<img src={item.image} alt="alt" />
			</Thumbnail>

			<PortfolioControl onClick={onChange}>
				<h3>{item.name}</h3>
				<p>{item.description}</p>
			</PortfolioControl>
		</PortfolioPreview>
		{isActive &&
			item.content.map((content, index) => (
				<div key={item.name + index}>
					<div>{content.blurb}</div>
					<p>{content.paragraph}</p>
				</div>
			))}
	</PortfolioItem>
);

const Portfolio = () => {
	const [active, setActive] = useState(null);

	return (
		<PortfolioList data-test="portfolio">
			{data.map(item => (
				<PortfolioEntry
					item={item}
					isActive={active === item.name}
					onChange={() => setActive(active !== item.name ? item.name : null)}
					key={item.name}
				/>
			))}
		</PortfolioList>
	);
};

export default Portfolio;
