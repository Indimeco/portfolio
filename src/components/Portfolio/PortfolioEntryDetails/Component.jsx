import React, { useEffect, useRef } from 'react';
import {
	PortfolioItemName,
	PortfolioItemDescription,
	PortfolioButton,
} from '../Component.style';
import {
	PortfolioItemContentHeader,
	PortfolioItemBlurb,
	PortfolioContentArea,
} from './Component.style';

const PortfolioContentGenerator = ({ type, value }) => {
	switch (type) {
		case 'paragraph':
			return (
				<PortfolioItemDescription data-testid="portfolio-item-content-desc">
					{value}
				</PortfolioItemDescription>
			);
		case 'blurb':
			return (
				<PortfolioItemBlurb data-testid="portfolio-item-content-blurb">
					{value}
				</PortfolioItemBlurb>
			);
		default:
			return null;
	}
};

const PortfolioEntryDetails = ({
	item: { name, description, content },
	onChange,
	...restProps
}) => {
	const headingRef = useRef(null);
	useEffect(() => {
		headingRef.current.scrollIntoView();
	}, []);
	return (
		<div>
			<PortfolioItemContentHeader ref={headingRef}>
				<PortfolioItemName>{name}</PortfolioItemName>
				<PortfolioButton onClick={() => onChange(null)}>
					Go back
				</PortfolioButton>
			</PortfolioItemContentHeader>
			<PortfolioContentArea data-testid="portfolio-item-content">
				<PortfolioItemDescription>{description}</PortfolioItemDescription>
				{content.map(({ type, value }, index) => (
					<React.Fragment key={name + index}>
						{PortfolioContentGenerator({ type, value })}
					</React.Fragment>
				))}
			</PortfolioContentArea>
		</div>
	);
};

export default PortfolioEntryDetails;
