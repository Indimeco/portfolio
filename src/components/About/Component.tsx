import React from 'react';

import { aboutId } from '../../content/about';

import Log from './Log';
import { AboutWrapper } from './Component.style';

export const About: React.FunctionComponent = () => (
	<AboutWrapper data-testid="about" id={aboutId}>
		<Log />
	</AboutWrapper>
);

export default About;
