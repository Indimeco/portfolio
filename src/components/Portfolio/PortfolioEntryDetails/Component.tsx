import React, { useEffect, useRef } from 'react';

import { PortfolioItemName, PortfolioItemDescription, PortfolioButton } from '../Component.style';
import { FadeOnScroll, useDyingLight } from '../../UI/utils';

import {
	PortfolioItemContentHeader,
	PortfolioItemBlurb,
	PortfolioItemLink,
	PortfolioContentArea,
	PortfolioContentFooter,
} from './Component.style';

const PortfolioLight: React.FC = ({ children }) => {
	const dyingLightProps = useDyingLight();
	return (
		<PortfolioItemBlurb data-testid="portfolio-item-content-blurb" {...dyingLightProps}>
			{children}
		</PortfolioItemBlurb>
	);
};

type ParagraphContent = {
	type: 'paragraph';
	value: string;
};
type BlurbContent = {
	type: 'blurb';
	value: string;
};
type LinkContent = {
	type: 'link';
	value: {
		href: string;
		text: string;
	};
};
type PortfolioContent = ParagraphContent | BlurbContent | LinkContent;

const PortfolioContentGenerator: React.FC<PortfolioContent> = (content) => {
	switch (content.type) {
		case 'paragraph':
			return (
				<PortfolioItemDescription data-testid="portfolio-item-content-desc">
					{content.value}
				</PortfolioItemDescription>
			);
		case 'blurb':
			return <PortfolioLight>{content.value}</PortfolioLight>;
		case 'link':
			return (
				<PortfolioItemLink as="a" href={content.value.href}>
					{content.value.text}
				</PortfolioItemLink>
			);
		default:
			return null;
	}
};

const PortfolioEntryDetails: React.FC<{
	item: { name: string; description: string; content: PortfolioContent[] };
	onChange: any;
	portfolioRef: any;
}> = ({ item: { name, content }, onChange, portfolioRef }) => {
	const headingRef = useRef<HTMLDivElement>(null);
	useEffect(() => {
		headingRef?.current?.scrollIntoView();
	}, []);
	return (
		<FadeOnScroll>
			<PortfolioItemContentHeader ref={headingRef}>
				<PortfolioItemName>{name}</PortfolioItemName>
				<PortfolioButton onClick={() => onChange(false)} data-ga={name}>
					Go back
				</PortfolioButton>
			</PortfolioItemContentHeader>
			<PortfolioContentArea data-testid="portfolio-item-content">
				{content.map((c, index) => (
					// eslint-disable-next-line
					<React.Fragment key={name + index}>{PortfolioContentGenerator(c)}</React.Fragment>
				))}
			</PortfolioContentArea>
			<PortfolioContentFooter>
				<PortfolioButton
					onClick={() => {
						onChange(false);
						portfolioRef.current.scrollIntoView();
					}}
					data-ga={name}
				>
					Go back
				</PortfolioButton>
			</PortfolioContentFooter>
		</FadeOnScroll>
	);
};

export default PortfolioEntryDetails;
