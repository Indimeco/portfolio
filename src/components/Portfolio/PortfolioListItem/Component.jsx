import React from 'react';
import {
	PortfolioItemName,
	PortfolioItemDescription,
	PortfolioButton,
} from '../Component.style';
import { PortfolioItem, PortfolioThumbnail } from './Component.style';

// TODO update CRA to 2.0 to use <> syntax
const PortfolioItemDetailsButton = ({ onChange }) => (
	<PortfolioButton onClick={onChange}>View details</PortfolioButton>
);

const PortfolioItemContentPreview = ({ name, description }) => (
	<React.Fragment>
		<PortfolioItemName>{name}</PortfolioItemName>
		<PortfolioItemDescription>{description}</PortfolioItemDescription>
	</React.Fragment>
);

const PortfolioListItem = ({ item, onChange }) => (
	<PortfolioItem data-testid="portfolio-list-item">
		<PortfolioThumbnail>
			<a href={item.href}>
				<img src={item.image} alt={item.alt} />
			</a>
		</PortfolioThumbnail>

		<div>
			<PortfolioItemContentPreview
				name={item.name}
				description={item.description}
			/>
			<PortfolioItemDetailsButton onChange={onChange} />
		</div>
	</PortfolioItem>
);

export default PortfolioListItem;
