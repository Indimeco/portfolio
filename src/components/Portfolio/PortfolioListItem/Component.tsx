import React, { HTMLAttributes } from 'react';

import { PortfolioItemName, PortfolioItemDescription, PortfolioButton } from '../Component.style';

import { PortfolioItem, PortfolioThumbnail } from './Component.style';

const PortfolioItemDetailsButton: React.FC<
	{ onChange: (e: React.MouseEvent) => void } & Omit<HTMLAttributes<HTMLButtonElement>, 'onChange'>
> = ({ onChange, ...restProps }) => (
	<PortfolioButton onClick={onChange} {...restProps}>
		View details
	</PortfolioButton>
);

const PortfolioItemContentPreview: React.FC<{ name: string; description: string }> = ({
	name,
	description,
}) => (
	<>
		<PortfolioItemName>{name}</PortfolioItemName>
		<PortfolioItemDescription>{description}</PortfolioItemDescription>
	</>
);

const PortfolioListItem: React.FC<{
	item: { name: string; description: string; image: string; alt: string };
	onChange: (e: React.MouseEvent) => void;
	index: number;
}> = ({ item, onChange, index }) => (
	<PortfolioItem data-testid="portfolio-list-item">
		<PortfolioThumbnail src={item.image} alt={item.alt} />

		<div>
			<PortfolioItemContentPreview name={item.name} description={item.description} />
			<PortfolioItemDetailsButton
				onChange={onChange}
				id={`portfolio-list-item-button-${index}`}
				data-ga={item.name}
			/>
		</div>
	</PortfolioItem>
);

export default PortfolioListItem;
