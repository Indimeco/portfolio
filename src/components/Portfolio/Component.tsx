import React, { useEffect, useRef, useState } from 'react';

import { portfolio } from '../../content';
import { FadeOnScroll } from '../UI/utils';

import PortfolioEntryDetails from './PortfolioEntryDetails';
import PortfolioListItem from './PortfolioListItem';
import { PortfolioView, PortfolioWrapper, PortfolioHeading } from './Component.style';

type Props = {
	title: string;
	setSectionBoundary: (pixelBoundary: number) => void;
};

const Portfolio: React.FC<Props> = ({ title, setSectionBoundary }) => {
	const [active, setActive] = useState<any>(null);
	const portfolioRef = useRef(null);
	const sectionBoundary = useRef<HTMLDivElement>(null);

	useEffect(() => {
		const node = sectionBoundary.current;
		if (node) {
			setSectionBoundary(node.offsetHeight + node.offsetTop);
		}
	}, [setSectionBoundary, sectionBoundary, portfolioRef]);
	return (
		<PortfolioView data-testid="portfolio" ref={sectionBoundary}>
			<PortfolioWrapper>
				<PortfolioHeading as="h2" ref={portfolioRef}>
					<FadeOnScroll>{title}</FadeOnScroll>
				</PortfolioHeading>
				{!active ? (
					<div>
						{portfolio.map((item, index) => (
							<FadeOnScroll key={item.name}>
								<PortfolioListItem item={item} onChange={() => setActive(item)} index={index} />
							</FadeOnScroll>
						))}
					</div>
				) : (
					<PortfolioEntryDetails item={active} onChange={setActive} portfolioRef={portfolioRef} />
				)}
			</PortfolioWrapper>
		</PortfolioView>
	);
};

export default Portfolio;
