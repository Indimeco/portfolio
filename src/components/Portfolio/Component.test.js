import React from 'react';
import renderer from 'react-test-renderer';
import Component from './Component';
import 'jest-styled-components';

describe('Portfolio', () => {
	it('renders correctly', () => {
		const tree = renderer.create(<Component />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
