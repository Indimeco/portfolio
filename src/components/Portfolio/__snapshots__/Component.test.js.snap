// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio renders correctly 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <button
      onClick={[Function]}
    >
      <h3>
        QRPG
      </h3>
      <p>
        QRPG is a Steam Workshop mod that provides UI and Data Management solutions for playing Role Playing Games within the physics sandbox game Tabletop Simulator.
      </p>
    </button>
    <img
      alt="alt"
      src="sample.png"
    />
  </div>
  <div
    className=""
  >
    <button
      onClick={[Function]}
    >
      <h3>
        My Little Charlotte
      </h3>
      <p>
        My Little Charlotte is a data scraping and data analysis tool, built using Puppeteer and Node JS.
      </p>
    </button>
    <img
      alt="alt"
      src="sample.png"
    />
  </div>
</div>
`;
